File: hard/218. The Skyline Problem/solution.cpp
class Solution {
public:
    vector<vector<int>> getSkyline(vector<vector<int>>& buildings) {
        vector<pair<int, int>> points;
        for (auto& b : buildings) {
            points.push_back({b[0], -b[2]});
            points.push_back({b[1], b[2]});
        }
        sort(points.begin(), points.end());
        
        multiset<int> heights = {0};
        vector<vector<int>> result;
        int prev = 0;
        
        for (auto& p : points) {
            int x = p.first, h = p.second;
            if (h < 0) {
                heights.insert(-h);
            } else {
                heights.erase(heights.find(h));
            }
            
            int curr = *heights.rbegin();
            if (curr != prev) {
                result.push_back({x, curr});
                prev = curr;
            }
        }
        return result;
    }
};

File: hard/218. The Skyline Problem/solution.rs
use std::collections::BTreeMap;

impl Solution {
    pub fn get_skyline(buildings: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut points = Vec::new();
        for b in &buildings {
            points.push((b[0], -b[2]));
            points.push((b[1], b[2]));
        }
        points.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1)));
        
        let mut heights = BTreeMap::new();
        heights.insert(0, 1);
        let mut result = Vec::new();
        let mut prev = 0;
        
        for (x, h) in points {
            if h < 0 {
                *heights.entry(-h).or_insert(0) += 1;
            } else {
                let entry = heights.entry(h).or_insert(0);
                *entry -= 1;
                if *entry == 0 {
                    heights.remove(&h);
                }
            }
            
            let curr = *heights.keys().rev().next().unwrap();
            if curr != prev {
                result.push(vec![x, curr]);
                prev = curr;
            }
        }
        result
    }
}